### DISCUSSION TITLE
Reconstruction of dumped model of inherited classes

### DISCUSSION THREAD

**Discussion by ahemwe** [pydantic/pydantic#8846] [Question] [answered]
Given this code, which is a simple representation of what I want to do in a fastapi solution.

``` python
from typing import List

from pydantic import BaseModel


class Building(BaseModel):
    size: int


class Plant(Building):
    workers: int


class MainClass(BaseModel):
    buildings: List[Building | Plant]


# Create instace
building = Building(size=10)
plant = Plant(size=20, workers=30)

initial_main = MainClass(buildings=[building, plant])

# Dump and restore
print(f"Original Model:         {initial_main}")
dump = initial_main.model_dump()
print(f"Dump of original model: {dump}")
print("")
new_main = MainClass(**dump)
print(f"Restored model:         {new_main}")

```

The output is 

``` python
Original Model:         buildings=[Building(size=10), Plant(size=20, workers=30)]
Dump of original model: {'buildings': [{'size': 10}, {'size': 20, 'workers': 30}]}

Restored model:         buildings=[Building(size=10), Building(size=20)]
```

meaning the instance of `Plant` is reduced to an instance of `Building` in reconstructed model.
When changing the order in type definition of the list to
`    buildings: List[Plant | Building]` (meaning `Plant` is the 1st type)
the result is correct:
``` python
Restored model:         buildings=[Building(size=10), Plant(size=20, workers=30)]
```

My questons:
- is this code meaningful?
- Is it expected that the order in the type definition is important?

Thanks

**Comment by GDYendell** [comment, marked as answer]
It is just returning the first valid match. It matches `Building` because the default is to ignore the extra `workers` field. If you added 

```python
class Building(BaseModel):
    model_config = ConfigDict(extra="forbid")
...
```

it will make validation for `{'buildings': [{'size': 10}, {'size': 20, 'workers': 30}]}` against `Building` fail and it will then try `Plant` and succeed.
